src/QirRuntime/lib/QIR/qirTypes.hpp

```diff
@@ -74,57 +74,87 @@ struct QirString
     QirString(std::string&& str);
 };
 
-/*==============================================================================
-    The types and methods, expected by QIR for tuples:
-    %TupleHeader = type { i32, i32 }
-    declare %TupleHeader* @__quantum__rt__tuple_create(i64)
-
-    Argument passed to __quantum__rt__tuple_create is the size (in bytes) of the tuple, including the header.
-    For example:
-    ; to calculate the size of a tuple pretend having an array of them and get
-    ; offset to the first element
-    %t1 = getelementptr { %TupleHeader, %Callable*, %Array* }, { %TupleHeader, %Callable*, %Array* }* null, i32 1
-    ; convert the offset to int and call __quantum__rt__tuple_create
-    %t2 = ptrtoint { %TupleHeader, %Callable*, %Array* }* %t1 to i64
-    %0 = call %TupleHeader* @__quantum__rt__tuple_create(i62 %t2)
-
-    Notice, that the TupleHeader is included into the Tuple's buffer.
-==============================================================================*/
+/*======================================================================================================================
+    Tuples are opaque to the runtime and the type of the data contained in them isn't (generally) known, thus, we use
+    char* to represent the tuples QIR operates with. However, we need to manage tuples' lifetime and in case of nested
+    controlled callables we also need to peek into the tuple's content. To do this we associate with each tuple's buffer
+    a header that contains the relevant data. The header immediately precedes the tuple's buffer in memeory when the
+    tuple is created.
+======================================================================================================================*/
+using PTuple = char*;
 struct QirTupleHeader
 {
     int refCount = 0;
-    int tupleSize = 0; // when creating the tuple, must be set to: size of this header + size of the tuple's data buffer
+    int tupleSize = 0; // when creating the tuple, must be set to the size of the tuple's data buffer
+
+    PTuple AsTuple()
+    {
+        return reinterpret_cast<PTuple>(this) + sizeof(QirTupleHeader);

```

Comment:
Is this arithmetic really correct?

Reply:
Currently QirTupleHeader is aligned but alignment is a concern. I'd love to add more protections here. Would you suggest any specific alignment annotations and static_asserts for this?

Reply:
Per Dmitry's suggestion, am using the flexible array member idiom now which should take care of alignment issues.